pkg/mimir/mimir.go

```diff
@@ -557,7 +557,10 @@ func (t *Mimir) Run() error {
 		return err
 	}
 
-	t.API.RegisterServiceMapHandler(http.HandlerFunc(t.servicesHandler))
+	// Some modules don't depend on the API, and therefore don't initialize it.
+	if t.API != nil {
+		t.API.RegisterServiceMapHandler(http.HandlerFunc(t.servicesHandler))
```

Comment:
Wouldn't that still panic on `:569` or `:571`?

Also, i'm not sure if it's desirable to make it possible to start Mimir without the API, what's the use case of that? The user would end up with some black-box process into which they'd have barely any insight. Instead of adding checks to test whether the API has been initialized we might want to consider to just make API a dependency of everything.

Reply:
> Wouldn't that still panic on :569 or :571?

No, because if `t.Ring` or `t.Ingester` are not nil, that means the ring and ingester have been initialized which means the API has been initialized because it is a dependency of both modules.

> Also, i'm not sure if it's desirable to make it possible to start Mimir without the API, what's the use case of that? The user would end up with some black-box process into which they'd have barely any insight. Instead of adding checks to test whether the API has been initialized we might want to consider to just make API a dependency of everything.

🤷🏻‍♂️ There are modules where the API is pretty much useless anyway, like `usage-stats`. This just seemed like the more "dskit-way" of doing things I guess. I don't really care though either way though.

Reply:
Also I think making api a dependency of server would cause a circular dependency error.

Reply:
I see there's a `testMode` variable. Maybe only allow non-user-visible targets if that's true?

Reply:
I don't think it's worthwhile to allow them at all. We can find another way to test if needed.